go/vt/vtgate/planbuilder/operators/route_planning.go

```diff
@@ -537,8 +537,30 @@ func pushJoinPredicates(ctx *plancontext.PlanningContext, exprs []sqlparser.Expr
 	}
 
 	for _, expr := range exprs {
-		AddPredicate(ctx, op, expr, true, newFilterSinglePredicate)
+		addJoinPredicate(ctx, op, expr)
 	}
 
 	return op
 }
+
+func addJoinPredicate(
+	ctx *plancontext.PlanningContext,
+	join *ApplyJoin,
+	expr sqlparser.Expr,
+) Operator {
+	deps := ctx.SemTable.RecursiveDeps(expr)
+	switch {
+	case deps.IsSolvedBy(TableID(join.LHS)):
```

Comment:
I created a new function instead of AddPredicate, because we need to store the join predicate even when it is pushed to only the right side. Basically the `join.JoinPredicates.addRight(exp)` call. Otherwise it is a simplified version of AddPredicate without the join and newfilter argument.